core/src/plugins/kubernetes/results-cache-base.ts

```diff
@@ -0,0 +1,207 @@
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+import type { PluginContext } from "../../plugin-context.js"
+import type { Log } from "../../logger/log-entry.js"
+import type { RunResult } from "../../plugin/base.js"
+import { runResultSchemaZod } from "../../plugin/base.js"
+import type { NamespaceStatus } from "../../types/namespace.js"
+import { namespaceStatusSchema } from "../../types/namespace.js"
+import type { RunAction } from "../../actions/run.js"
+import type { TestAction } from "../../actions/test.js"
+import { hashSync } from "hasha"
+import { Memoize } from "typescript-memoize"
+import type { AnyZodObject, z } from "zod"
+import { deline } from "../../util/string.js"
+import type { ContainerRunAction, ContainerTestAction } from "../container/config.js"
+import type { KubernetesRunAction, KubernetesTestAction } from "./kubernetes-type/config.js"
+import type { HelmPodRunAction, HelmPodTestAction } from "./helm/config.js"
+import { renderZodError } from "../../config/zod.js"
+import type { JsonObject } from "type-fest"
+import { GardenError } from "../../exceptions.js"
+
+export type CacheableAction = RunAction | TestAction
+
+export type CacheableRunAction = ContainerRunAction | KubernetesRunAction | HelmPodRunAction
+export type CacheableTestAction = ContainerTestAction | KubernetesTestAction | HelmPodTestAction
+
+export type SchemaVersion = `v${number}`
+
+/**
+ * Increment the current result schema format version
+ * if the result format changes breaking backwards-compatibility,
+ * e.g. serialization format.
+ */
+export const currentResultSchemaVersion: SchemaVersion = "v1"
+
+export const kubernetesCacheEntrySchema = runResultSchemaZod.extend({
+  namespaceStatus: namespaceStatusSchema.required(),
+})
+
+export type KubernetesCacheEntrySchema = typeof kubernetesCacheEntrySchema
+
+export function composeKubernetesCacheEntry({
+  result,
+  namespaceStatus,
+}: {
+  result: RunResult
+  namespaceStatus: NamespaceStatus
+}): z.input<KubernetesCacheEntrySchema> {
+  return {
+    ...result,
+    namespaceStatus,
+  }
+}
+
+export interface LoadResultParams<A extends CacheableAction> {
+  ctx: PluginContext
+  log: Log
+  action: A
+}
+
+export type ClearResultParams<A extends CacheableAction> = LoadResultParams<A>
+
+export interface StoreResultParams<A extends CacheableAction, R> {
+  ctx: PluginContext
+  log: Log
+  action: A
+  result: R
+}
+
+export class StructuredCacheKey {
+  private readonly projectName: string
+  private readonly actionKind: string
+  private readonly actionName: string
+  private readonly actionType: string
+  private readonly actionVersion: string
+
+  constructor({ ctx, action }: { ctx: PluginContext; action: CacheableAction }) {
+    this.projectName = ctx.projectName
+    this.actionKind = action.kind
+    this.actionName = action.name
```

Comment:
```suggestion
    this.actionKind = action.kind
```
I think we don't need these two. Why should we enforce a cache miss if the name changes?

Reply:
Right, simplified this in 059e47bc369ef8ae3c8c9bbaa8952191af6626e7.